# Homework 4
## Ziyan Hu

## Question 1
1. Navigate to data.cityofnewyork.us (April 23, 2023). Click the export button to download
the data as a .csv file. Hint: It should be downloaded with a name like
2018 Central Park Squirrel Census - Squirrel Data.csv.
a. Read in the file to the skeleton code .ipynb. For full credit, code the path argument
in a way that shows you’ve created a unique folder for this assignment, and, that a
special sub-folder separates the .ipynb from the .csv.
b. This dataset has 30+ columns. That is a bit too much to keep track of in your
head. Please create a new dataframe called “focused squirrel df”, that only has these
twelve columns: [“X”, “Y”, “Unique Squirrel ID”, “Shift”, “Date”, “Age”, “Primary
Fur Color”, “Highlight Fur Color”, “Above Ground Sighter Measurement”, “Kuks”,
“Quaas”, “Tail flags”]. Are all rows in this new table complete? If not, how many
rows have missing data?
c. Lets focus on the “Unique Squirrel ID” column. Using this column, tell us the date
(day, month) on which the most productive chronological sequence of squirrel sightings
took place. How many squirrels were seen? Hint: Check the download link to read
how it was constructed.
d. Create a new dataframe with just the squirrels from that sequence. To answer if
this survey might be ‘double counting’, please use .groupby([[...]])[“Unique Squirrel
ID”].count() to answer how many squirrels had the same age, primary fur color, and
secondary fur color. Do you think double counting took place?
e. Kuks vs Quaas. Now we can answer the truly important question; which form of
squirrel communication is more common. Use a print statement to display the number
of observations of Kuks, and another for number of Quaas observations. Please use
all the data in “focused squirrel df”. Finally, return the column name that was most
common. Do so without typing the name out as a “string"

In [1]:
# 1a. read the csv in a sub folder
import pandas as pd
import numpy as np
path = "data/2018_Central_Park_Squirrel_Census_-_Squirrel_Data_20231012.csv"
squirrel_data = pd.read_csv(path)
squirrel_data.head()

X          Y Unique Squirrel ID Hectare Shift      Date  \
0 -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1 -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2 -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3 -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4 -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   

   Hectare Squirrel Number    Age Primary Fur Color Highlight Fur Color  ...  \
0                        3    NaN               NaN                 NaN  ...   
1                        4    NaN               NaN                 NaN  ...   
2                        8    NaN              Gray                 NaN  ...   
3                       14  Adult              Gray                 NaN  ...   
4                        5  Adult              Gray            Cinnamon  ...   

    Kuks  Quaas  Moans Tail flags Tail twitches  Approaches  Indifferent  \
0  False  False  False      False         False       False        False   
1  False  False  False      False         False       False        False   
2  False  False  False      False         False       False        False   
3  False  False  False      False         False       False        False   
4  False  False  False      False         False       False        False   

   Runs from  Other Interactions                                    Lat/Long  
0      False                 NaN  POINT (-73.9561344937861 40.7940823884086)  
1      False                 NaN  POINT (-73.9688574691102 40.7837825208444)  
2      False                 NaN  POINT (-73.97428114848522 40.775533619083)  
3       True                 NaN  POINT (-73.9596413903948 40.7903128889029)  
4      False                 NaN  POINT (-73.9702676472613 40.7762126854894)  

[5 rows x 31 columns]

In [93]:
# 1b. create a new dataframe
focused_squirrel_df = squirrel_data[['X', 'Y', 'Unique Squirrel ID', 'Shift', 'Date', 'Age', 'Primary Fur Color','Highlight Fur Color', 'Above Ground Sighter Measurement', 'Kuks','Quaas', 'Tail flags']]
focused_squirrel_df.isnull().sum()

X                                      0
Y                                      0
Unique Squirrel ID                     0
Shift                                  0
Date                                   0
Age                                  121
Primary Fur Color                     55
Highlight Fur Color                 1086
Above Ground Sighter Measurement     114
Kuks                                   0
Quaas                                  0
Tail flags                             0
dtype: int64

The rows in this table is not complete. From the above table, we can see that Age has 121 missing data, Primary Fun Color has 55, Highlight Fur Color has 1086, Above Ground Sighter Measurement has 114.

In [94]:
# 1c.
# Extracting date and month from "Unique Squirrel ID"
focused_squirrel_df = focused_squirrel_df.copy()
focused_squirrel_df['Month'] = focused_squirrel_df['Unique Squirrel ID'].str.extract(r'-(\d{2})\d{2}-')
focused_squirrel_df['Day'] = focused_squirrel_df['Unique Squirrel ID'].str.extract(r'-(\d{2})$')
# Combine the month and day columns into a single 'Date' column
focused_squirrel_df['Extracted Date'] = focused_squirrel_df['Month'] + "/" + focused_squirrel_df['Day']
# Group by the extracted date and count the number of sightings
counts = focused_squirrel_df['Extracted Date'].value_counts()
# Get the date with the highest number of sightings
date = counts.idxmax()
highest_count = counts.max()
date, highest_count

('10/01', 614)

The date on which the most productive chronological sequence of squirrel sightingstook place is 10/01 and there are 614 squirrels were seen.

In [95]:
# 1d.
# creare a new data frame
squirrel = squirrel_data[['Unique Squirrel ID','Age','Primary Fur Color','Highlight Fur Color']]
squirrel.groupby('Age')['Unique Squirrel ID'].count()

Age
?              4
Adult       2568
Juvenile     330
Name: Unique Squirrel ID, dtype: int64

In [96]:
# group by primary fur color
squirrel.groupby('Primary Fur Color')['Unique Squirrel ID'].count()

Primary Fur Color
Black        103
Cinnamon     392
Gray        2473
Name: Unique Squirrel ID, dtype: int64

In [97]:
# group by secondary fur color
squirrel.groupby('Highlight Fur Color')['Unique Squirrel ID'].count()

Highlight Fur Color
Black                      34
Black, Cinnamon             9
Black, Cinnamon, White     32
Black, White               10
Cinnamon                  767
Cinnamon, White           268
Gray                      170
Gray, Black                 3
Gray, White                59
White                     585
Name: Unique Squirrel ID, dtype: int64

There is double counting took place.

In [98]:
# 1e.
# Count and print the number of occurrences of "Kuks" and "Quaas"
kuks_count = focused_squirrel_df['Kuks'].sum()
quaas_count = focused_squirrel_df['Quaas'].sum()

# Displaying the count of "Kuks" and "Quaas" using print statements
print(f"Number of observations of Kuks: {kuks_count}")
print(f"Number of observations of Quaas: {quaas_count}")

# Determine which communication method is more common
most_common_communication = "Kuks" if kuks_count > quaas_count else "Quaas"
most_common_communication

Number of observations of Kuks: 102
Number of observations of Quaas: 50


'Kuks'

## Question 2
Quantitative squirrelnalysis:
a. Three columns in this data have data that aren’t categorical. Which best fit that description?
b. For the applicable set of observations, what was the average above ground sighter measurement? Compute two more averages, splitting the first into average in the “AM” hours, and average in the “PM” hours.
c. Intro to ‘geo-coding’: The “X” and “Y” columns are associated with longitude and latitude. Create a new column which calculates the distance of each observation to [40.765, -73.954]. Euclidean distance is acceptable here, but in other geo-coding applications, the curvature of earth’s surface does matter. Which squirrel was seen farthest from our lecture hall?
d. Compute the average squirrel position in the “AM” hours, then do the same for “PM” hours. Which direction does the population tend to migrate on the average day cycle? Lastly, list factors which might have biased this conclusion.
e. Correlation analysis: Pandas includes some mathematical operations that can workon dataframes. Create a new dataframe that only has the columns “Y” and “Above Ground Sighter Measurement”. Use pandas.pydata.org/docs to calculate the correlation between how north in the park a squirrel was observed and the height at which it was seen in a tree. What direction is the association

1a.
There are columns in this data have data that aren't categorical. That is 'X', 'Y' and 'Above Ground Sighter Measurement'. 'X' and 'Y' represent geographical coordinates, which are numerical and continuous. 'Above Ground Sighter Measurement' has numerical values.

In [102]:
# 1b.
# Convert "Above Ground Sighter Measurement" to numeric, setting errors='coerce' to handle non-numeric values
focused_squirrel_df["Above Ground Sighter Measurement"] = pd.to_numeric(focused_squirrel_df["Above Ground Sighter Measurement"], errors='coerce')

# Calculate the overall average above ground sighter measurement
avg_measurement = focused_squirrel_df["Above Ground Sighter Measurement"].mean()
print(f"the average above ground sighter measurements: {avg_measurement}")

# Calculate the average above ground sighter measurement for AM and PM shifts
avg_measurement_am = focused_squirrel_df[focused_squirrel_df['Shift'] == 'AM']['Above Ground Sighter Measurement'].mean()
avg_measurement_pm = focused_squirrel_df[focused_squirrel_df['Shift'] == 'PM']['Above Ground Sighter Measurement'].mean()
print(f"the average above ground sighter measurements in the AM hours: {avg_measurement_am}")
print(f"the average above ground sighter measurements in the PM hours: {avg_measurement_pm}")

the average above ground sighter measurements: 15.210592686002522
the average above ground sighter measurements in the AM hours: 16.58254716981132
the average above ground sighter measurements in the PM hours: 13.634146341463415


In [103]:
# 1c.
# coordinates of the reference point [latitude, longitude]
lecture_hall = [40.765, -73.954]

# calculate the distance and create a new column
focused_squirrel_df = focused_squirrel_df.copy()
focused_squirrel_df['Distance'] = np.sqrt(
    (focused_squirrel_df['X']-lecture_hall[0])**2 +
    (focused_squirrel_df['Y']-lecture_hall[1])**2
)

# find the squirrel that was seen farthest from the reference point
max_distance = focused_squirrel_df['Distance'].max()
squirrel_id = focused_squirrel_df.loc[focused_squirrel_df['Distance'].idxmax(), 'Unique Squirrel ID']
print(f"The squirrel with ID {squirrel_id} has the max distance from our lecture hall, which is {avg_measurement_pm}")

The squirrel with ID 42A-AM-1019-01 has the max distance from our lecture hall, which is 13.634146341463415


In [104]:
# 1d.
# compute the average position for squirrels in AM and PM
avg_position_am = [focused_squirrel_df[focused_squirrel_df['Shift'] == 'AM']['X'].mean(),
                      focused_squirrel_df[focused_squirrel_df['Shift'] == 'AM']['Y'].mean()]
avg_position_pm = [focused_squirrel_df[focused_squirrel_df['Shift'] == 'PM']['X'].mean(),
                   focused_squirrel_df[focused_squirrel_df['Shift'] == 'PM']['Y'].mean()]
(avg_position_am,avg_position_pm)

([-73.96750477893433, 40.78051006555636],
 [-73.96692591313666, 40.78112862780465])

Squirrels tend to move northeast from AM to PM, as the latitude increases and the longitude increases (becomes less negative). The factors that affect the migration might be food resources, weather conditions, human activity.

In [105]:
# 1e.
# create a new data frame that only contain 2 columns
df = focused_squirrel_df[['Y','Above Ground Sighter Measurement']].copy()
# Ensure "Above Ground Sighter Measurement" is numeric
df["Above Ground Sighter Measurement"] = pd.to_numeric(df["Above Ground Sighter Measurement"], errors='coerce')
# calculate the correlation between Y and Above Ground Sighter Measurement
correlation = df["Y"].corr(df["Above Ground Sighter Measurement"])
correlation

0.06254168873204949

A correlation of 0.063 indicates a weak positive association between the latitude at which a squirrel is observed (how far north in the park) and the height at which it was seen in a tree. This suggests that, to a very small degree, as one variable increases, the other tends to increase as well.